tests/framework/helpers/FileHelperTest.php

```diff
@@ -790,4 +790,59 @@ public function testCopyDirectoryExclude()
             $this->assertStringEqualsFile($fileName, $content, 'Incorrect file content!');
         }
     }
+
+    /**
+     * @see https://github.com/yiisoft/yii2/issues/9669
+     *
+     * @depends testCopyDirectory
+     * @depends testFindFiles
+     */
+    public function testCopyDirectoryEmptyDirectories()
+    {
+        $srcDirName = 'test_empty_src_dir';
+        $this->createFileStructure([
+            $srcDirName => [
+                'dir1' => [
+                    'file1.txt' => 'file1',
+                    'file2.txt' => 'file2',
+                ],
+                'dir2' => [
+                    'file1.log' => 'file1',
+                    'file2.log' => 'file2',
+                ],
+                'dir3' => [],
+            ],
+        ]);
+
+        $basePath = $this->testFilePath;
+        $srcDirName = $basePath . DIRECTORY_SEPARATOR . $srcDirName;
+
+        // copy with empty directories
+        $dstDirName = $basePath . DIRECTORY_SEPARATOR . 'test_empty_dst_dir';
+        FileHelper::copyDirectory($srcDirName, $dstDirName, ['only' => ['*.txt'], 'copyEmptyDirectories' => true]);
+
+        $this->assertFileExists($dstDirName, 'Destination directory does not exist!');
+        $copiedFiles = FileHelper::findFiles($dstDirName);
+        $this->assertCount(2, $copiedFiles, 'wrong files count copied');
+
+        $this->assertFileExists($dstDirName . DIRECTORY_SEPARATOR . 'dir1');
+        $this->assertFileExists($dstDirName . DIRECTORY_SEPARATOR . 'dir1' . DIRECTORY_SEPARATOR . 'file1.txt');
+        $this->assertFileExists($dstDirName . DIRECTORY_SEPARATOR . 'dir1' . DIRECTORY_SEPARATOR . 'file2.txt');
+        $this->assertFileExists($dstDirName . DIRECTORY_SEPARATOR . 'dir2');
+        $this->assertFileExists($dstDirName . DIRECTORY_SEPARATOR . 'dir3');
+
```

Comment:
this test should be split  in more tests with clear 3 parts:
1) test setup
2) execution
3) assertions

This will make the test a lot easier to read & to debug.